packages/vscode-docker-registries/src/clients/GenericRegistryV2/GenericRegistryV2DataProvider.ts

```diff
@@ -0,0 +1,153 @@
+import { RegistryWizardSecretPromptStep, RegistryWizardUsernamePromptStep } from '../../wizard/RegistryWizardPromptStep';
+import { isContextValueRegistryItem } from '../../contracts/RegistryItem';
+
+const GenericV2StorageKey = 'GenericV2ContainerRegistry';
+const TrackedRegistriesKey = `${GenericV2StorageKey}.TrackedRegistries`;
+
+interface GenericV2RegistryRoot extends V2RegistryRoot {
+    readonly additionalContextValues: ['genericRegistryV2Root'];
+}
+
+export interface GenericV2Registry extends V2Registry {
+    readonly additionalContextValues: ['genericRegistryV2Registry'];
+}
+
+interface GenericV2RegistryTag extends V2Tag {
+    readonly additionalContextValues: ['genericRegistryV2Tag'];
+}
+
+export function isGenericV2Registry(item: unknown): item is GenericV2Registry {
+    return isContextValueRegistryItem(item) && item.additionalContextValues?.includes('genericRegistryV2Registry') === true;
+}
+
+export class GenericRegistryV2DataProvider extends RegistryV2DataProvider {
+    public readonly id = 'vscode-docker.genericRegistryV2DataProvider';
+    public readonly label = vscode.l10n.t('Generic Registry V2');
+    public readonly description: undefined;
+    public readonly iconPath = new vscode.ThemeIcon('link');
+
+    private readonly authenticationProviders = new Map<string, BasicOAuthProvider>(); // The tree items are too short-lived to store the associated auth provider so keep a cache
+
+    public constructor(private readonly extensionContext: vscode.ExtensionContext) {
+        super();
+    }
+
+    public async onConnect(): Promise<void> {
+        await this.addTrackedRegistry();
+    }
+
+    public getRoot(): GenericV2RegistryRoot {
+        return {
+            parent: undefined,
+            label: this.label,
+            type: 'commonroot',
+            iconPath: this.iconPath,
+            additionalContextValues: ['genericRegistryV2Root'],
+        };
+    }
+
+    public async getRegistries(root: GenericV2RegistryRoot | GenericV2Registry): Promise<GenericV2Registry[]> {
+        const trackedRegistryStrings = this.extensionContext.globalState.get<string[]>(TrackedRegistriesKey, []);
+        const trackedRegistries = trackedRegistryStrings.map(r => vscode.Uri.parse(r));
+
+        return trackedRegistries.map(r => {
+            return {
+                label: r.toString(),
+                parent: root,
+                type: 'commonregistry',
+                additionalContextValues: ['genericRegistryV2Registry'],
+                baseUrl: r
+            };
+        });
+    }
+
+    public async getTags(repository: V2Repository): Promise<GenericV2RegistryTag[]> {
+        const tags = await super.getTags(repository);
+        const tagsWithAdditionalContext: GenericV2RegistryTag[] = tags.map(tag => ({
+            ...tag,
+            additionalContextValues: ['genericRegistryV2Tag']
+        }));
+
+        return tagsWithAdditionalContext;
+    }

```

Comment:
What purpose does this additional context value serve?

Reply:
After we make the context value change in delete image, this is no longer needed

Reply:
https://github.com/microsoft/vscode-docker/pull/4026#discussion_r1313331522 should simplify this as well